sympy/core/symbol.py

```diff
@@ -42,6 +42,15 @@ def __getnewargs__(self):
     def _hashable_content(self):
         return (self.name,)
 
+    def __eq__(self, other):
+        if isinstance(other, str):
```

Comment:
It shouldn't be necessary to define `__eq__` and `__hash__` as this is handled in the superclass using `_hashable_content`.

I'm not sure why you say that hash isn't implemented for `Str`. On master I get:
```julia
In [1]: from sympy.core.symbol import Str                                                                                                                     

In [2]: hash(Str('x'))                                                                                                                                        
Out[2]: 7290937937051250587
```

Reply:
If I override the `__eq__` as in the PR but don't override `__hash__`, I get the follwing:

```python
from sympy.core.symbol import Str
hash(Str('x'))
# TypeError: unhashable type: 'Str'
```

Reply:
Sorry, I should not have said "not implemented".

Reply:
Also, `__eq__` is needed, otherwise `Str('x') != 'x'`.

Reply:
I don't think that we should allow `Str('x') == 'x'`. If that is needed then maybe `Str` isn't the right thing.

Reply:
Maybe the user should supply symbols for the state space.

Reply:
I managed to get it working now without overriding. I personally feel that its pretty un-intuitive that `Str('x') != 'x'`. If we have a user playing around with the state space they may find:
```python
from sympy.stats import *

X = DiscreteMarkovChain('X', ['Sunny', 'Cloudy', 'Rainy'])
# this could happen in a user's loop somewhere
print(X.state_space[0] == 'Sunny')  # False
```
I guess that will be the price to pay for string state spaces. 
No current working API has changed at all though.

Reply:
Maybe it should be:
```python
from sympy import symbols
from sympy.stats import *

Sunny, Cloudy, Rainy = states = symbols('Sunny, Cloudy, Rainy')

X = DiscreteMarkovChain('X', states)
# this could happen in a user's loop somewhere
print(X.state_space[0] == Sunny) # True
```
Really I see `Str` as just being a kind of categorical symbol. It's a symbol but with the understanding that it can not be used in e.g. arithmetic or mathematical functions.

Reply:
I think that also works with this PR. This PR basically adds any kind of hashable state. It's just that `str` is the only one with the nack of becoming something completely different when sympified.

Reply:
This is why it's better to use `_sympify` rather than `sympify` or otherwise to sympify with `strict=True`.

Reply:
It's better to reject strings than evaluate them the way that sympify does.